lib/vector-core/src/stream/driver.rs

```diff
@@ -0,0 +1,189 @@
+        other
+            .partial_cmp(self)
+            .expect("PendingAcknowledgement should always return a valid comparison")
+    }
+}
+
+/// Drives the interaction between a stream of items and a service which processes them
+/// asynchronously.
+///
+/// `Driver`, as a high-level, facilitates taking items from an arbitrary `Stream` and pushing them
+/// through a `Service`, spawning each call to the service so that work can be run concurrently,
+/// managing waiting for the service to be ready before processing more items, and so on.
+///
+/// Additionally, `Driver` handles two event-specific facilities: finalization and acknowledgement.
+///
+/// This capability is parameterized so any implementation which can define how to interpret the
+/// response for each request, as well as define how many events a request is compromised of, can be
+/// used with `Driver`.
+pub struct Driver<St, Svc> {
+    input: St,
+    service: Svc,
+    acker: Acker,
+}
+
+impl<St, Svc> Driver<St, Svc> {
+    pub fn new(input: St, service: Svc, acker: Acker) -> Self {
+        Self {
+            input,
+            service,
+            acker,
+        }
+    }
+}
+
+impl<St, Svc> Driver<St, Svc>
+where
+    St: Stream,
+    St::Item: Ackable + Finalizable,
+    Svc: Service<St::Item>,
+    Svc::Error: fmt::Debug + 'static,
+    Svc::Future: Send + 'static,
+    Svc::Response: AsRef<EventStatus>,
+{
+    /// Runs the driver until the input stream is exhausted.
+    ///
+    /// All in-flight calls to the provided `service` will also be completed before `run` returns.
+    ///
+    /// # Errors
+    ///
+    /// No errors are currently returned.  The return type is purely to simplify caller code, but may
+    /// return an error for a legitimate reason in the future.
+    pub async fn run(self) -> Result<(), ()> {
+        let mut in_flight = FuturesUnordered::new();
+        let mut pending_acks = BinaryHeap::new();
+        let mut seq_head: u64 = 0;
+        let mut seq_tail: u64 = 0;
+
+        let Self {
+            input,
+            mut service,
+            acker,
+        } = self;
+
+        pin!(input);
+
+        loop {
+            select! {
+                // Using `biased` ensures we check the branches in the order they're written, and
+                // the way they're ordered is to ensure that we're reacting to completed requests as
+                // soon as possible to acknowledge them and make room for more requests to be processed.
+                biased;
+
+                // One of our service calls has completed.
+                Some(result) = in_flight.next() => {
+                    // Rebind so we can annotate the type, otherwise the compiler is inexplicably angry.
+                    let result: Result<(u64, usize), JoinError> = result;
+                    match result {
+                        Ok((seq_no, ack_size)) => {
+                            trace!(message = "Sending request.", seq_no, ack_size);
```

Comment:
Is this message accurate? It seems like we've already sent the request and gotten a response at this point.

Reply:
Oh! Good catch.